### ISSUE TITLE
안녕하세요. GAIL 코드 구현에 있어 한 가지 질문이 있습니다.

### ISSUE THREAD

**Issue by bhappy10** [chagmgang/gail#1]
안녕하세요. 차금강님 코드를 통해 GAIL을 구현하는 중에 질문이 있어 글을 남깁니다.

제가 GAIL 논문을 보았을 때 E[log(agent에서 나온 데이터)] - E[log(1 - expert 데이터)] 로 알고리즘을 이해했습니다.

그리고, 차금강님의 discriminator 코드를 보면 34 ~ 43 line에서 아래와 같이 코드를 구현해주셨습니다.

 with tf.variable_scope('loss'):
                **loss_expert = tf.reduce_mean(tf.log(tf.clip_by_value(prob_1, 0.01, 1)))
                loss_agent = tf.reduce_mean(tf.log(tf.clip_by_value(1 - prob_2, 0.01, 1)))**
                loss = loss_expert + loss_agent
                loss = -loss
                tf.summary.scalar('discriminator', loss)
optimizer = tf.train.AdamOptimizer()
self.train_op = optimizer.minimize(loss)
**self.rewards = tf.log(tf.clip_by_value(prob_2, 1e-10, 1))**

제가 차금강님의 코드를 보고 이해한 것이 맞다면 차금강님께서는
E[log(expert 데이터)] - E[log(1 - agent에서 나온 데이터)]
로 구현하신 듯 하여 여기에 어떤 이유가 있는지 여쭤보고 싶습니다.

그리고 제가 차금강님의 코드를 아래처럼 조금 바꿔서 수렴이 좀 더 빠르게 되었는데 이 부분에 문제점이 없는지도 여쭤보고 싶습니다.
with tf.variable_scope('loss'):
                **loss_expert = tf.reduce_mean(tf.log(tf.clip_by_value(1-prob_1, 0.01, 1)))
                loss_agent = tf.reduce_mean(tf.log(tf.clip_by_value(prob_2, 0.01, 1)))**
                l = loss_expert + loss_agent
                self.loss = -l
                tf.summary.scalar('discriminator', self.loss)
optimizer = tf.train.AdamOptimizer()
self.train_op = optimizer.minimize(self.loss)
**self.rewards = -tf.log(tf.clip_by_value(prob_2, 1e-10, 1))**

마지막으로, 아래의 첫 번째 그래프는 차금강님의 코드로 환경에서 주는 reward(적색), discriminator에서 주는 reward(청색)을 나타낸 것입니다. 두 번째 그래프는 제가 조금 수정한 코드의 그래프 입니다. 

![graph](https://user-images.githubusercontent.com/17981048/47264627-8f6caa80-d555-11e8-918e-bc1ed5e3c350.png)

![graph8](https://user-images.githubusercontent.com/17981048/47264736-5e8d7500-d557-11e8-8b1e-713bb881d57c.png)

**Comment by chagmgang** [comment]
관리를 하지 않는 레포에 이렇게 이슈를 남겨주셔서 감사합니다.
논문대로 구현이 되지는 않았지만 정확하게 논문대로 구현하는 것을 목표로 하신다면
고치신 것이 맞는 것 같습니다.
논문에서는 E[log(agent에서 나온 데이터)] + E[log(1 - expert 데이터)] 로 표현하고 있지만
이것을 GAN의 개념으로 본다면 
무엇이 진짜인지 가짜인지 구별하는 문제가 기존 논문의 방향이었다면
제가 짠 코드는 무엇이 가짜인지 진짜인지 구별하는 문제로 바뀐 것으로 보입니다.
논문의 내용적으로는 문제가 되지 않지만
논문대로 구현을 하여 더 좋은 성능을 나타낸다면 그것이 맞는 구현인것 같습니다.
PR 날려주시면 반영하도록 하겠습니다.
